File: internal/modules/ratelimit/bucket.lua
-- Token bucket rate limiter, executado atomicamente dentro do Redis.
--
-- KEYS[1] = chave do hash que representa o bucket (ex: "bucket:user1")
-- ARGV[1] = maxTokens    (capacidade máxima do bucket)
-- ARGV[2] = refillRate   (tokens repostos por segundo)
-- ARGV[3] = requestCost  (quantos tokens este request consome)
-- ARGV[4] = now          (timestamp atual em segundos, fornecido pelo cliente Go)
--
-- Retorna 1 se o request foi permitido, 0 se foi negado.

local maxTokens = tonumber(ARGV[1])
local refillRate = tonumber(ARGV[2])
local requestCost = tonumber(ARGV[3])
local now = tonumber(ARGV[4])

-- HMGET busca os dois campos de uma vez. Se a chave não existir, o Redis
-- devolve `false` (não `nil`) para cada campo ausente -- convenção do
-- Redis-Lua, diferente do Lua puro.
local vals = redis.call('HMGET', KEYS[1], 'tokens', 'timestamp')

local currentTokens
local lastTimestamp

if vals[1] == false then
    -- bucket novo: começa cheio, igual você decidiu na versão in-memory
    currentTokens = maxTokens
    lastTimestamp = now
else
    -- tudo que vem do Redis chega como STRING, mesmo números -- por isso
    -- o tonumber(...) aqui é obrigatório, senão a aritmética abaixo quebra
    currentTokens = tonumber(vals[1])
    lastTimestamp = tonumber(vals[2])
end

-- mesma proteção contra relógio andando pra trás que você já fez no
-- gerador de ID -- aqui, em vez de rejeitar com erro, simplesmente não
-- concede refill negativo (mais tolerante, já que o pior caso aqui é
-- "um pouco menos generoso", não uma colisão de dados)
local elapsed = now - lastTimestamp
if elapsed < 0 then
    elapsed = 0
end

local tokensToAdd = elapsed * refillRate
currentTokens = math.min(currentTokens + tokensToAdd, maxTokens)

local allowed = 0
if currentTokens >= requestCost then
    currentTokens = currentTokens - requestCost
    allowed = 1
end

redis.call('HMSET', KEYS[1], 'tokens', currentTokens, 'timestamp', now)

-- TTL: tempo que levaria pra esse bucket encher de novo do zero, mais uma
-- margem. Assim, buckets de chaves que pararam de ser usadas não ocupam
-- memória do Redis pra sempre.
local ttl = 3600
if refillRate > 0 then
    ttl = math.ceil(maxTokens / refillRate) + 60
end
redis.call('EXPIRE', KEYS[1], ttl)

return allowed

File: internal/modules/ratelimit/bucket.go
package ratelimit

import (
	"context"
	_ "embed"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"
)

//go:embed bucket.lua
var tokenBucketScript string


const requestCost = 1.0



type Limiter interface {
	Allow(ctx context.Context, key string) (bool, error)
}

type RedisLimiter struct {
	client     *redis.Client
	namespace  string
	maxTokens  float64
	refillRate float64
	script     *redis.Script
}

func NewRedisLimiter(client *redis.Client, namespace string, maxTokens float64, refillRate float64) *RedisLimiter {
	script := redis.NewScript(tokenBucketScript)
	return &RedisLimiter{
		client:     client,
		namespace:  namespace,
		maxTokens:  maxTokens,
		refillRate: refillRate,
		script:     script,
	}
}

func (l *RedisLimiter) Allow(ctx context.Context, key string) (bool, error) {
	now := float64(time.Now().Unix())
	combinedKey := "bucket:" + l.namespace + ":" + key

	result, err := l.script.Run(ctx, l.client, []string{combinedKey}, l.maxTokens, l.refillRate, requestCost, now).Int()
	if err != nil {
		return false, err
	}

	return result == 1, nil
}

type bucket struct {
	currentTokens float64
	lastTimestamp time.Time
}

type InMemoryLimiter struct {
	mu           sync.Mutex
	buckets      map[string]*bucket
	maxTokens    float64
	refillRate   float64 // tokens por segundo

}

func NewInMemoryLimiter(maxTokens float64, refillRate float64) *InMemoryLimiter {
	return &InMemoryLimiter{
		buckets: make(map[string]*bucket),
		maxTokens: maxTokens,
		refillRate: refillRate,
	}
}

func (l *InMemoryLimiter) Allow(ctx context.Context, key string) (bool, error) {
	l.mu.Lock()
	defer l.mu.Unlock()

	b, exists := l.buckets[key]
	if !exists {
		
		newBucket := &bucket{
			currentTokens: l.maxTokens - requestCost,
			lastTimestamp: time.Now(),
		}
		l.buckets[key] = newBucket
		return true, nil
	}
	now := time.Now()
	elapsed := now.Sub(b.lastTimestamp)          
	tokensToAdd := elapsed.Seconds() * l.refillRate

	b.currentTokens = min(b.currentTokens+tokensToAdd, l.maxTokens)
	b.lastTimestamp = now

	if b.currentTokens >= requestCost {
		b.currentTokens -= requestCost
		return true, nil
	}

	return false, nil
}